File: clients/js/src/haveapi/i18n.js
var I18n = {
	DEFAULT_LOCALE: 'en',
	DEFAULT_LANGUAGE_HEADER: 'Accept-Language',

	translate: function(language, key, values) {
		var message = this.lookup(this.localeFor(language), key) ||
			this.lookup(this.DEFAULT_LOCALE, key) ||
			key;

		values = values || {};

		for (var name in values) {
			if (!values.hasOwnProperty(name))
				continue;

			message = message.replace(new RegExp('%\\{'+ name +'\\}', 'g'), String(values[name]));
		}

		return message;
	},

	requestHeaders: function(language, languageHeader) {
		var ret = {};

		if (language === undefined || language === null || String(language) === '')
			return ret;

		languageHeader = languageHeader || this.DEFAULT_LANGUAGE_HEADER;
		this.assertHeaderName(languageHeader);
		this.assertHeaderValue(language);

		ret[languageHeader] = String(language);
		return ret;
	},

	localeFor: function(language) {
		if (language === undefined || language === null || String(language) === '')
			return this.DEFAULT_LOCALE;

		var tags = this.parseAcceptLanguage(String(language));

		for (var i = 0; i < tags.length; i++) {
			var locale = this.normalizeLocale(tags[i]);

			if (locale && I18nMessages.hasOwnProperty(locale))
				return locale;
		}

		return this.DEFAULT_LOCALE;
	},

	assertHeaderName: function(name) {
		if (typeof name !== 'string' || !name.match(/^[!#$%&'*+.^_`|~0-9A-Za-z-]+$/))
			throw new Error('Invalid language HTTP header name');
	},

	assertHeaderValue: function(value) {
		if (typeof value !== 'string' && typeof value !== 'number' && typeof value !== 'boolean')
			throw new Error('Invalid language HTTP header value');

		if (String(value).match(/[\x00\r\n]/))
			throw new Error('Invalid language HTTP header value');
	},

	lookup: function(locale, key) {
		var data = I18nMessages[locale];
		var parts = String(key).replace(/^haveapi_client\./, '').split('.');

		for (var i = 0; i < parts.length; i++) {
			if (!data || !data.hasOwnProperty(parts[i]))
				return null;

			data = data[parts[i]];
		}

		return typeof data === 'string' ? data : null;
	},

	parseAcceptLanguage: function(header) {
		var ret = [];
		var parts = header.split(',');

		for (var i = 0; i < parts.length; i++) {
			var tokens = parts[i].split(';');
			var tag = tokens.shift().trim();
			var q = 1.0;

			for (var j = 0; j < tokens.length; j++) {
				var qTokens = tokens[j].split('=');

				if (qTokens[0].trim() == 'q')
					q = parseFloat(qTokens[1]);
			}

			if (tag !== '' && q > 0)
				ret.push([tag, q]);
		}

		ret.sort(function(a, b) {
			return b[1] - a[1];
		});

		return ret.map(function(v) {
			return v[0];
		});
	},

	normalizeLocale: function(tag) {
		var normalized = String(tag).trim().replace(/_/g, '-').replace(/\..*$/, '').toLowerCase();

		if (normalized === '')
			return null;

		return normalized.split('-', 1)[0];
	}
};

File: clients/ruby/lib/haveapi/client/i18n.rb
require 'yaml'

module HaveAPI::Client
  module I18n
    DEFAULT_LOCALE = :en
    DEFAULT_LANGUAGE_HEADER = 'Accept-Language'.freeze
    HEADER_NAME = /\A[A-Za-z0-9!#$%&'*+\-.^_`|~]+\z/
    HEADER_VALUE_UNSAFE = /[\x00\r\n]/

    class << self
      def t(language, key, values = {})
        message = lookup(locale_for(language), key) || lookup(DEFAULT_LOCALE, key) || key.to_s

        interpolate(message, values)
      end

      def request_headers(language, language_header = DEFAULT_LANGUAGE_HEADER)
        return {} if language.nil? || language.to_s.empty?

        assert_header_name!(language_header)
        assert_header_value!(language)

        { language_header => language.to_s }
      end

      def locale_for(language)
        return DEFAULT_LOCALE if language.nil? || language.to_s.empty?

        parse_accept_language(language).each do |tag|
          normalized = normalize_locale(tag)
          return normalized if normalized && messages.has_key?(normalized)
        end

        DEFAULT_LOCALE
      end

      def assert_header_name!(header)
        return if header.is_a?(String) && header.match?(HEADER_NAME)

        raise ArgumentError, "invalid language HTTP header name: #{header.inspect}"
      end

      def assert_header_value!(value)
        return if value.is_a?(String) && !value.match?(HEADER_VALUE_UNSAFE)

        raise ArgumentError, "invalid language HTTP header value: #{value.inspect}"
      end

      private

      def lookup(locale, key)
        key.to_s.sub(/\Ahaveapi_client\./, '').split('.').reduce(messages[locale]) do |data, part|
          break unless data.is_a?(Hash)

          data[part]
        end
      end

      def interpolate(message, values)
        values.reduce(message.dup) do |ret, (key, value)|
          ret.gsub("%{#{key}}", value.to_s)
        end
      end

      def parse_accept_language(header)
        header.to_s.split(',').filter_map do |part|
          tag, *params = part.split(';')
          q = 1.0

          params.each do |param|
            name, value = param.split('=', 2).map(&:strip)
            q = value.to_f if name == 'q'
          end

          next if tag.strip.empty? || q <= 0

          [tag.strip, q]
        end.sort_by { |(_tag, q)| -q }.map(&:first)
      end

      def normalize_locale(tag)
        normalized = tag.to_s.strip.tr('_', '-').sub(/\..*\z/, '').downcase
        return if normalized.empty?

        primary = normalized.split('-', 2).first.to_sym
        primary if messages.has_key?(primary)
      end

      def messages
        @messages ||= begin
          dir = File.expand_path('locales', __dir__)

          Dir[File.join(dir, '*.yml')].to_h do |file|
            locale = File.basename(file, '.yml').to_sym
            data = YAML.safe_load_file(file, aliases: true) || {}

            [locale, data.fetch(locale.to_s).fetch('haveapi_client')]
          end
        end
      end
    end
  end
end
